File: apps/irontide/frontend/src/net/protocol.ts
/**
 * Network protocol types — mirrors the Rust server's ServerMessage/ClientMessage enums.
 */

// Server -> Client messages
export interface RoomCreated { type: 'RoomCreated'; room_code: string; player_id: number; }
export interface RoomJoined { type: 'RoomJoined'; player_id: number; }
export interface GameStart { type: 'GameStart'; seed: number; map_seed: number; player_count: number; }
export interface TurnCommands { type: 'TurnCommands'; tick: number; commands: PlayerCommands[]; }
export interface DesyncDetected { type: 'DesyncDetected'; tick: number; }
export interface PlayerDisconnected { type: 'PlayerDisconnected'; player_id: number; }
export interface ServerError { type: 'Error'; message: string; }

export interface PlayerCommands {
  player_id: number;
  commands_json: string;
  checksum: string | null;
}

export type ServerMessage =
  | RoomCreated
  | RoomJoined
  | GameStart
  | TurnCommands
  | DesyncDetected
  | PlayerDisconnected
  | ServerError;

// Client -> Server messages
export interface CreateRoomMsg { type: 'CreateRoom'; }
export interface JoinRoomMsg { type: 'JoinRoom'; room_code: string; }
export interface GameCommandsMsg { type: 'GameCommands'; tick: number; commands_json: string; checksum: string | null; }

export type ClientMessage = CreateRoomMsg | JoinRoomMsg | GameCommandsMsg;

File: apps/irontide/engine/crates/irontide-server/src/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ClientMessage {
    CreateRoom,
    JoinRoom { room_code: String },
    GameCommands { tick: u32, commands_json: String, checksum: Option<String> },
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ServerMessage {
    RoomCreated { room_code: String, player_id: u8 },
    RoomJoined { player_id: u8 },
    GameStart { seed: u64, map_seed: u64, player_count: u8 },
    TurnCommands { tick: u32, commands: Vec<PlayerCommands> },
    DesyncDetected { tick: u32 },
    PlayerDisconnected { player_id: u8 },
    Error { message: String },
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PlayerCommands {
    pub player_id: u8,
    pub commands_json: String,
    pub checksum: Option<String>,
}
